From 112e5aebb1fd985e8ef7eb0de83e349d70d801e8 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sat, 18 Jul 2026 02:21:38 -0400
Subject: [PATCH 199/268] HACK: winewayland: Implement OpenGL throttling.

Works around driver limitations.
---
 dlls/win32u/opengl.c                   | 11 ++++++-
 dlls/winewayland.drv/opengl.c          | 41 ++++++++++++++++++++++++--
 dlls/winewayland.drv/wayland_surface.c |  6 +++-
 dlls/winewayland.drv/waylanddrv.h      |  3 ++
 4 files changed, 57 insertions(+), 4 deletions(-)

diff --git a/dlls/win32u/opengl.c b/dlls/win32u/opengl.c
index bb3a2e0eae0..2f14eec6e85 100644
--- a/dlls/win32u/opengl.c
+++ b/dlls/win32u/opengl.c
@@ -1406,10 +1406,19 @@ static BOOL egldrv_make_current( struct opengl_drawable *draw, struct opengl_dra
 {
     const struct opengl_funcs *funcs = &display_funcs;
     const struct egl_platform *egl = &display_egl;
+    BOOL ret;
 
     TRACE( "draw %s, read %s, context %p\n", debugstr_opengl_drawable( draw ), debugstr_opengl_drawable( read ), context );
 
-    return funcs->p_eglMakeCurrent( egl->display, context ? draw->surface : EGL_NO_SURFACE, context ? read->surface : EGL_NO_SURFACE, context );
+    ret = funcs->p_eglMakeCurrent( egl->display, context ? draw->surface : EGL_NO_SURFACE, context ? read->surface : EGL_NO_SURFACE, context );
+    if (ret && context && draw->surface != EGL_NO_SURFACE && egl->display && egl->type == EGL_PLATFORM_WAYLAND_KHR)
+    {
+        /* HACK: set swap interval to 0 for the wayland platform,
+         * winewayland implements swap interval=1 instead of the GPU driver. */
+        ERR("HACK: Using swap interval 0\n");
+        funcs->p_eglSwapInterval( egl->display, 0 );
+    }
+    return ret;
 }
 
 static void egldrv_pbuffer_destroy( struct opengl_drawable *drawable )
diff --git a/dlls/winewayland.drv/opengl.c b/dlls/winewayland.drv/opengl.c
index 8a47f3d3e4b..80b0ec716ef 100644
--- a/dlls/winewayland.drv/opengl.c
+++ b/dlls/winewayland.drv/opengl.c
@@ -116,6 +116,8 @@ static BOOL wayland_opengl_surface_create(HWND hwnd, BOOL raw, int format, struc
     gl->base.buffer_map[GL_FRONT - GL_FRONT_LEFT] = GL_BACK;
     gl->base.buffer_map[GL_FRONT_AND_BACK - GL_FRONT_LEFT] = GL_BACK;
 
+    NtCreateEvent(&client->throttle, EVENT_ALL_ACCESS, NULL, SynchronizationEvent, TRUE);
+
     if (!(gl->wl_egl_window = wl_egl_window_create(client->wl_surface, rect.right, rect.bottom))) goto err;
     if (!(gl->base.surface = funcs->p_eglCreateWindowSurface(egl->display, config, gl->wl_egl_window, attribs))) goto err;
     set_client_surface(hwnd, client);
@@ -145,18 +147,53 @@ static void wayland_drawable_flush(struct opengl_drawable *base, UINT flags)
 
     TRACE("drawable %s, flags %#x\n", debugstr_opengl_drawable(base), flags);
 
-    if (flags & GL_FLUSH_INTERVAL) funcs->p_eglSwapInterval(egl->display, abs(base->interval));
-
     /* Since context_flush is called from operations that may latch the native size,
      * perform any pending resizes before calling them. */
     if (flags & GL_FLUSH_UPDATED) wayland_gl_drawable_sync_size(gl);
 }
 
+static void wayland_gl_frame_done(void *data, struct wl_callback *wl_callback, uint32_t callback_data)
+{
+    struct client_surface *client = data;
+    struct wayland_client_surface *surface = impl_from_client_surface(client);
+
+    if (surface->wl_callback != wl_callback) return;
+
+    TRACE("hwnd=%p\n", client->hwnd);
+
+    surface->wl_callback = NULL;
+    if (surface->throttle) NtSetEvent(surface->throttle, NULL);
+}
+
+static const struct wl_callback_listener gl_throttle_listener =
+{
+    wayland_gl_frame_done,
+};
+
 static BOOL wayland_drawable_swap(struct opengl_drawable *base)
 {
     struct wayland_gl_drawable *gl = impl_from_opengl_drawable(base);
+    struct wayland_client_surface *surface = impl_from_client_surface(base->client);
 
     client_surface_present(base->client);
+    if (base->interval && surface->toplevel)
+    {
+        /* 1 second, relative */
+        const LARGE_INTEGER timeout = { .QuadPart = -10000 * 1000 };
+        NTSTATUS status;
+
+        if (surface->throttle)
+        {
+            status = NtWaitForSingleObject(surface->throttle, FALSE, &timeout);
+            if (status == STATUS_TIMEOUT) WARN("Present timed out!\n");
+        }
+       /* Currently, EGL is implemented using frame callbacks, so replicate that here.
+        * although it would be better to use fifo + commit timing
+        * (that way swap interval > 1 can be implemented),
+        * that's not something easily done from the wine side. */
+        surface->wl_callback = wl_surface_frame(surface->wl_surface);
+        wl_callback_add_listener(surface->wl_callback, &gl_throttle_listener, base->client);
+    }
     funcs->p_eglSwapBuffers(egl->display, gl->base.surface);
 
     return TRUE;
diff --git a/dlls/winewayland.drv/wayland_surface.c b/dlls/winewayland.drv/wayland_surface.c
index 257b91356ec..9abfa5c435e 100644
--- a/dlls/winewayland.drv/wayland_surface.c
+++ b/dlls/winewayland.drv/wayland_surface.c
@@ -1374,7 +1374,7 @@ POINT map_point_from_surface(struct wayland_surface *surface, POINT point)
     return point;
 }
 
-static struct wayland_client_surface *impl_from_client_surface(struct client_surface *client)
+struct wayland_client_surface *impl_from_client_surface(struct client_surface *client)
 {
     return CONTAINING_RECORD(client, struct wayland_client_surface, client);
 }
@@ -1385,6 +1385,10 @@ static void wayland_client_surface_destroy(struct client_surface *client)
 
     TRACE("%s\n", debugstr_client_surface(client));
 
+    if (surface->wl_callback)
+        wl_callback_destroy(surface->wl_callback);
+    if (surface->throttle)
+        NtClose(surface->throttle);
     if (surface->wp_color_management_surface_v1)
         wp_color_management_surface_v1_destroy(surface->wp_color_management_surface_v1);
     if (surface->wp_content_type_v1)
diff --git a/dlls/winewayland.drv/waylanddrv.h b/dlls/winewayland.drv/waylanddrv.h
index 2df573ae285..3d0c39d1650 100644
--- a/dlls/winewayland.drv/waylanddrv.h
+++ b/dlls/winewayland.drv/waylanddrv.h
@@ -359,6 +359,8 @@ struct wayland_client_surface
 {
     struct client_surface client;
     HWND toplevel;
+    HANDLE throttle;
+    struct wl_callback *wl_callback;
     struct wl_surface *wl_surface;
     struct wl_subsurface *wl_subsurface;
     struct wp_color_management_surface_v1 *wp_color_management_surface_v1;
@@ -466,6 +468,7 @@ struct wayland_client_surface *wayland_client_surface_create(HWND hwnd);
 void wayland_client_surface_attach(struct wayland_client_surface *client, HWND toplevel);
 void wayland_client_surface_attach_image_description(struct client_surface *client,
                                                      struct wp_image_description_v1 *image_desc);
+struct wayland_client_surface *impl_from_client_surface(struct client_surface *client);
 void wayland_surface_set_title(struct wayland_surface *surface, LPCWSTR title);
 void wayland_surface_assign_icon(struct wayland_surface *surface);
 void wayland_surface_set_icon_buffer(struct wayland_surface *surface, UINT type, const ICONINFO *ii);
-- 
2.55.0

